services/proxy/pkg/userroles/oidcroles.go

```diff
@@ -0,0 +1,227 @@
+package userroles
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	cs3 "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	types "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	revactx "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/owncloud/ocis/v2/ocis-pkg/middleware"
+	settingssvc "github.com/owncloud/ocis/v2/protogen/gen/ocis/services/settings/v0"
+	"go-micro.dev/v4/metadata"
+)
+
+type oidcRoleAssigner struct {
+	Options
+}
+
+// NewOIDCRoleAssigner returns an implemenation of the UserRoleAssigner interface
+func NewOIDCRoleAssigner(opts ...Option) UserRoleAssigner {
+	opt := Options{}
+	for _, o := range opts {
+		o(&opt)
+	}
+
+	return oidcRoleAssigner{
+		Options: opt,
+	}
+}
+
+// UpdateUserRoleAssignment assigns the role "User" to the supplied user. Unless the user
+// already has a different role assigned.
+func (ra oidcRoleAssigner) UpdateUserRoleAssignment(ctx context.Context, user *cs3.User, claims map[string]interface{}) (*cs3.User, error) {
+	logger := ra.logger.SubloggerWithRequestID(ctx).With().Str("userid", user.GetId().GetOpaqueId()).Logger()
+	claimValueToRoleID, err := ra.oidcClaimvaluesToRoleIDs()
+	if err != nil {
+		logger.Error().Err(err).Msg("Error mapping claims to roles ids")
+		return nil, err
+	}
+
+	roleIDsFromClaim := make([]string, 0, 1)
+	claimRoles, ok := claims[ra.rolesClaim].([]interface{})
+	if !ok {
+		logger.Error().Err(err).Str("rolesClaim", ra.rolesClaim).Msg("No roles in user claims")
+		return nil, err
+	}
+	logger.Debug().Str("rolesClaim", ra.rolesClaim).Interface("rolesInClaim", claims[ra.rolesClaim]).Msg("got roles in claim")
+	for _, cri := range claimRoles {
+		cr, ok := cri.(string)
+		if !ok {
+			err := errors.New("invalid role in claims")
+			logger.Error().Err(err).Interface("claimValue", cri).Msg("Is not a valid string.")
+			return nil, err
+		}
+		id, ok := claimValueToRoleID[cr]
+		if !ok {
+			logger.Debug().Str("role", cr).Msg("No mapping for claim role. Skipped.")
+			continue
+		}
+		roleIDsFromClaim = append(roleIDsFromClaim, id)
+	}
+	logger.Debug().Interface("roleIDs", roleIDsFromClaim).Msg("Mapped claim roles to roleids")
+
+	switch len(roleIDsFromClaim) {
+	default:
+		err := errors.New("too many roles found in claims")
+		logger.Error().Err(err).Msg("Only one role per user is allowed.")
+		return nil, err
+	case 0:
+		err := errors.New("no role in claim, maps to a ocis role")
+		logger.Error().Err(err).Msg("")
+		return nil, err
+	case 1:
+		// exactly one mapping. This is right
+		break
```

Comment:
Curious: The `break` is needed as the linter won't allow having an empty branch on the `switch`? Or is there another reason?

Reply:
Hm. I think that this was just my C habits breaking though. I am not even sure the linter would complain about this.